### PR TITLE
Implement a lint to warn about unused macro rules

### LINKED ISSUE #73576: Warn about unused macro arms.
`rustc` now warns about unused macros. With the following code:

```rust
macro_rules! used {
    () => {};
}

macro_rules! unused { // <- This should be cleaned up.
    () => {};
}

fn main() {
    used!();
} 
```
we get the warning:
```
warning: unused macro definition
 --> src/main.rs:5:1
  |
5 | / macro_rules! unused {
6 | |     () => {};
7 | | }
  | |_^
  |
  = note: `#[warn(unused_macros)]` on by default
```
This is really helpful while cleaning up a large meta-programming mess.

However, it does not warn about unused macro *arms*, so the following:

```rust
macro_rules! used_macro {
    (used_arm) => {};
    (unused_arm) => {}; // <-- This should be cleaned up.
}

fn main() {
    used_macro!(used_arm);
} 
```
yields no warning yet.

Following [#34938](https://github.com/rust-lang/rust/issues/34938#issuecomment-646693357), I open this issue for discussing this feature. Is this something desirable to have `rustc` warn about unused macro arms?

### PR DESCRIPTION
**Pull request by est31**
This implements a new lint to warn about unused macro rules (arms/matchers), similar to the `unused_macros` lint added by #41907 that warns about entire macros.

```rust
macro_rules! unused_empty {
    (hello) => { println!("Hello, world!") };
    () => { println!("empty") }; //~ ERROR: 2nd rule of macro `unused_empty` is never used
}

fn main() {
    unused_empty!(hello);
}
```

Builds upon #96149 and #96156.

Fixes #73576

### CODE REVIEW

File: compiler/rustc_resolve/src/macros.rs
Comment (fee1-dead) [inline compiler/rustc_resolve/src/macros.rs:918]: I'm not sure if this is helpful. Can it just be `this case is never used` instead, since we highlight the arm already?

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (fee1-dead) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:546]: Can this just be the left hand sides? Ideally we don't want to highlight a large code block. Perhaps you should add a few tests that have a rhs that spans over many lines?

File: library/core/src/internal_macros.rs
Comment (fee1-dead) [inline library/core/src/internal_macros.rs:248]: Should this actually be removed? I think it could be useful.

File: library/proc_macro/src/quote.rs
Comment (fee1-dead) [inline library/proc_macro/src/quote.rs:15]: Again, I'm not sure if this should be removed.

File: src/test/ui/lint/unused/unused-macro-rules.rs
Comment (fee1-dead) [inline src/test/ui/lint/unused/unused-macro-rules.rs:31]: Does this lint avoid linting at public/exported macros? You might have to add those as test cases.

File: compiler/rustc_resolve/src/macros.rs
Comment (est31) [reply compiler/rustc_resolve/src/macros.rs:918]: Arms might look pretty similar to each other. Not all editors show the line number. IMO it's better to give two ways to find the thing the lint complains about.

File: src/test/ui/lint/unused/unused-macro-rules.rs
Comment (est31) [reply src/test/ui/lint/unused/unused-macro-rules.rs:31]: Yes, it avoids that. The lint shares the logic with the `unused_macros` lint for this part, but a refactor might change that, so I guess a few tests won't hurt.

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (est31) [reply compiler/rustc_expand/src/mbe/macro_rules.rs]: Oh indeed, for the `unused_macros` lint a similar change was conducted in #90761. For the tests, it should already be tested due to the `^^^^`s in the snippets. E.g. it prints this:

```
error: 2nd rule of macro `num` is never used
  --> $DIR/unused-macro-rules-decl.rs:9:5
   |
LL |     (two) => { 2 },
   |     ^^^^^

```

To be on the safe side though, I've changed one test each to have an rhs with multiple lines.

File: library/core/src/internal_macros.rs
Comment (est31) [reply library/core/src/internal_macros.rs:248]: This change is part of #96149. Let's see what the reviewers of that PR say.

File: library/proc_macro/src/quote.rs
Comment (est31) [reply library/proc_macro/src/quote.rs:15]: This change is part of #96149. Let's see what the reviewers of that PR say.

File: compiler/rustc_resolve/src/macros.rs
Comment (fee1-dead) [reply compiler/rustc_resolve/src/macros.rs:918]: > Not all editors show the line number. IMO it's better to give two ways to find the thing the lint complains about.

If some editors don't show line numbers, they should at least highlight the span that had the lint. There are a lot of errors and lints that don't make sense if there is no context given (line number, column number, etc.), and this would be one of them.

Comment (iago-lito) [reply compiler/rustc_resolve/src/macros.rs:918]: Side note : `ordinalize(21)` yields `1st` and not `21st`, which I'm not sure is desired.

Comment (iago-lito) [reply compiler/rustc_resolve/src/macros.rs:918]: Btw which arm are you referring to as the `1st`? The one with index `0` or the one with index `1`?

Comment (est31) [reply compiler/rustc_resolve/src/macros.rs]: > I've noticed that ordinalize(21) yields 1st and not 21st, which I'm not sure is desired.

Good catch! Indeed it was not desired. I have fixed it. I've wrote some tests but they didn't run because they were disabled in Cargo.toml, now they are enabled.

> Btw which arm are you referring to as the 1st? The one with index 0 or the one with index 1?

The one with index 0, there is a `+1` where `ordinalize` is being called. You can do some bikeshed on this question I guess, but note that it can be changed in the future. Personally I'm fine with either way.

Its ordinal (as well as line/column numbers) is the only watertight way to specify a macro arm. There is no rule that says that matchers, or even entire arms, have to be different from each other. For example, this is allowed without a warning currently:

```Rust
macro_rules! t {
    () => { println!("Hi") };
    () => { println!("Hi") };
    () => { println!("Hi") };
}
```

Ideally, there would be unused warnings for the second and third arm.

In practice, I admit this is rarer but it occurs often that macro arms look very similar to each other and you have to squint to figure out the differences.

**Comment by est31** [comment]
#96149 is merged now. r? @fee1-dead

**Review by fee1-dead — APPROVED** [review]
LGTM! Thanks

**Comment by est31** [comment]
@petrochenkov idk there are different lints when a function is never used (dead_code) vs when one of its parameters is never used (unused_variables).

Furthermore, if the lints are separate, it's easier to express "unused rules are okay, but if the entire macro is unused, please warn me" intent. Especially as right now you can't attach allow attributes to the single rules but have to do it on the entire macro.

On the other hand, if someone does `#[allow(unused_macros)]` on a macro and then some code starts to use it but only a couple of the rules, they might not like it that they now get *another* lint. For them though, the `#[allow(unused)]` directive exists, which is more powerful and shorter to type.

File: src/test/ui/rust-2018/async-ident.rs
Comment (petrochenkov) [inline src/test/ui/rust-2018/async-ident.rs:1]: Non-`run-pass` tests allow the whole `unused` lint group, so for them adding `allow(unused_macro_rules)` shouldn't be necessary.

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (petrochenkov) [inline compiler/rustc_lint_defs/src/builtin.rs:779]: I'd add some wording saying that the rules in question are inside the macro, and give an explicit example.
Otherwise it's confusable with whole `macro_rules` items.

File: compiler/rustc_resolve/src/macros.rs
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:27]: Nit: could you avoid multi-line imports?

**Comment by petrochenkov** [comment]
(I didn't finish reviewing yet.)

File: compiler/rustc_resolve/src/macros.rs
Comment (est31) [reply compiler/rustc_resolve/src/macros.rs:27]: How should I do it? If I move it to the same line, rustfmt makes it multiline again. I could refer to `UNUSED_MACRO_RULES` via the full path, but that would be inconsistent with the other lints.

File: src/test/ui/rust-2018/async-ident.rs
Comment (est31) [reply src/test/ui/rust-2018/async-ident.rs:1]: The issue is that this test also has rustfix enabled, and once that has ran, there are no warnings allowed. I've tried removing the `unused_macro_rules` allow again, it made the test fail.

File: compiler/rustc_expand/src/base.rs
Comment (petrochenkov) [inline compiler/rustc_expand/src/base.rs:716]: So my main concern was that this data doesn't belong to `SyntaxExtension` in general, only to declarative macros.

Could you return `rule_spans` separately from `compile_(declarative_)macro` instead of merging it to `SyntaxExtension`?

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:159]:
```suggestion
    node_id: NodeId,
```
Could you also replace the previously existing `is_local` field with this, instead of adding a new field?

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:302]: Only need to record use if `id != DUMMY_NODE_ID`, otherwise it's a macro decoded from some other crate.

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:543]: Only need to collect spans if `id != DUMMY_NODE_ID`, otherwise it's a macro decoded from some other crate.

File: compiler/rustc_resolve/src/lib.rs
Comment (petrochenkov) [inline compiler/rustc_resolve/src/lib.rs:976]:
```suggestion
    unused_macro_rules: FxHashMap<(LocalDefId, usize), (Ident, Span)>,
```
- `NodeId` and `LocalDefId` here are the same thing so only one needs to be kept, the `NodeId` for linting can be obtained with `self.def_id_to_node_id[def_id]`
- The rule span can also be kept here if it's no longer kept in `SyntaxExtension`

File: compiler/rustc_resolve/src/macros.rs
Comment (petrochenkov) [reply compiler/rustc_resolve/src/macros.rs:27]: By splitting it into multiple imports.

File: compiler/rustc_resolve/src/macros.rs
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:908]: Could you move this function to the `diagnostics.rs` module together with the tests for it?

File: compiler/rustc_resolve/src/macros.rs
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:320]:
```suggestion
        if let def_id = self.local_def_id(node_id) {
```
The def-id must exist if `DUMMY_NODE_ID`s were filtered away previously.

**Comment by petrochenkov** [comment]
r=me after squashing review commits into main commits.
@rustbot author

**Comment by est31** [comment]
> r=me after squashing review commits into main commits.

Done!

r? @petrochenkov

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Comment by GuillaumeGomez** [comment]
It's working (too) well and failed in https://github.com/rust-lang/rust/pull/96775. :)

@bors r-

**Comment by est31** [comment]
The fix seems to have fixed it:

Screenshot_20220508_223342

So it's good to go, or blocked on https://github.com/rust-lang/stdarch/pull/1305 , depending on policy.

**Comment by petrochenkov** [comment]
Does stdarch submodule points to some non-master branch now?
I haven't seen the written explicitly as a policy, but submodules aren't supposed to point to "temporary" commits.
So this is blocked until https://github.com/rust-lang/stdarch/pull/1305 lands on stdarch master.

**Comment by est31** [comment]
Another go, now `test-various` is fixed too. I suggest going for another try, and if there are still failures, just allowing the `unused_macro_rules` lint for all of `stdarch`.

**Comment by ComputerDruid** [comment]
Hello! I'm dropping by to give a little experience report on this lint. This week, I'm in charge of updating the nightly toolchain used by the Fuchsia project, which sets `-D warnings` by default, so new warnings actually prevent toolchain rolls. That's not a problem, we can just [allow new lints as necessary], but it does mean I got an overview of a bunch of instances of this lint from the Fuchsia codebase. Note that I didn't write the code in question, nor have I necessarily even met the people who have. If it were helpful, I could probably find someone to explain what something does, though.

[allow new lints as necessary]: https://fuchsia-review.googlesource.com/c/fuchsia/+/679506

Here's a change showing all the macro rules this lint wanted to remove: https://fxrev.dev/c/679455/1

One can't be removed because the macro rules are sometimes used based on `#[cfg(target_arch = "x86_64")]`: https://fxrev.dev/c/fuchsia/+/679455/1/src/connectivity/lib/internet-checksum/src/lib.rs . I'm not sure if it's possible to `cfg`-out macro rules (like you would do with a function that was only used in one config) but I'd guess not? It probably will be possible to refactor that code in a way that doesn't trigger the lint, though.

A couple of rules which solely exist to allow trailing commas [1](https://fuchsia-review.googlesource.com/c/fuchsia/+/679455/1/src/connectivity/network/netstack3/core/src/macros.rs) [2](https://fuchsia-review.googlesource.com/c/fuchsia/+/679455/1/src/proc/bin/starnix/fs/fs_node.rs), were flagged as unused. I don't think these should be removed though, because allowing trailing commas is good even if no one is using that right now. It's possible that this lint should detect and allow that, or suggest some alternative syntax to unify the macro rules if there's a good way to allow trailing commas in one rule.

One rule is used to give a better compile error when an impossible condition is used: https://fxrev.dev/c/fuchsia/+/679455/1/src/lib/mundane/src/boringssl/wrapper.rs , so it being unused is expected.

I don't understand why the rules from https://fxrev.dev/c/fuchsia/+/679455/1/src/connectivity/lowpan/lib/spinel_pack/src/primitives.rs are unused, but I would hesitate to remove the first one since if code in the future were to use it, it would result in a change in behavior rather than a compile failure (since the `u8` case would fall through to the `$self: ty` case).

The rest seem like they were included to try to be comprehensive of a matrix of possibilities, but some combinations of arguments end up going unused, for example: https://fxrev.dev/c/fuchsia/+/679455/1/src/connectivity/network/netstack3/core/src/ip/gmp/mod.rs . I'm unsure whether the lint is helpful here; it doesn't feel quite the same as an unused function, more like an "unused combination of arguments to a function". So this lint feels a bit noisy to me in this case.

I haven't dug into any of these especially deeply, but the lint did seem to have a lower signal:noise ratio than I expected, so I thought I would share. Hope these examples are helpful!

**Comment by est31** [comment]
@ComputerDruid thanks for the experience report!

> I'm not sure if it's possible to cfg-out macro rules (like you would do with a function that was only used in one config) but I'd guess not? 

Currently it's not possible to cfg-out single macro arms, no (or to allow the lint for single arms only). The [recommended solution](https://github.com/rust-lang/rust/issues/73576#issuecomment-647344401) is to allow the lint at the top of the macro declaration via `#[allow(unknown_lints, unused_macro_arms)]`.

>  It's possible that this lint should detect and allow that, or suggest some alternative syntax to unify the macro rules if there's a good way to allow trailing commas in one rule.

There is a good way to allow trailing commas, indeed, via the `?` kleene operator. The RFC [contains an example for trailing commas](https://rust-lang.github.io/rfcs/2298-macro-at-most-once-rep.html#trailing-commas). In theory it is possible to implement heuristics like "does it call the macro itself inside with the rule input repeated and just a , added to the end", which probably catch 70% or more of cases, and that maybe a clippy lint to suggest the `?` kleene operator *in the general setting* (even if there is no trailing comma) can utilize.

**Comment by est31** [comment]
> One rule is used to give a better compile error when an impossible condition is used: https://fxrev.dev/c/fuchsia/+/679455/1/src/lib/mundane/src/boringssl/wrapper.rs , so it being unused is expected.

I'm wondering about making the lint not fire in this instance. On one hand, it feels a bit niche/surprising behavior to me, but on the other hand, there are few "parse time" lints that could have possible interactions with `compile_error!`, so it's expected to be niche. 

Furthermore, the *very point* of `compile_error!` is to be not expanded during a successful compilation, as in, it exists to be "unused". A direct invocation of `compile_error!` is easy to detect, but it could be hidden through indirection, which is quite hard to unwrangle so I think not a good idea. I think that supporting direct invocations first would be a good start.

> but I would hesitate to remove the first one since if code in the future were to use it, it would result in a change in behavior rather than a compile failure (since the u8 case would fall through to the $self: ty case).

Yeah I guess in that case it's better to add an `#[allow(...)]`. For the different combinations of arguments use case, it probably depends on your personal taste. I lean towards removing stuff as it is still preserved in git history, but shrug.

Thanks for writing all of this up, @ComputerDruid !

-------

As any possible changes to the lint to make it more permissive (like the ones I discussed above) won't make it into the 1.62 release, which is the first release to feature the lint, it might be a good idea to turn the lint to `allow` by default until they are implemented. The second advantage is that it makes the introduction easier, as in, you can avoid allowing the `unknown_lints` lint in more cases (depending on the Rust compiler versions you as a developer care about). I wonder what others think.

**Comment by fee1-dead** [comment]
I wonder if this is more suitable for clippy. If this kind of lint has a high chance of linting something that it is not supposed to, then it might just be too annoying to be in rustc.

**Comment by est31** [comment]
That's an interesting suggestion. Rust lints have a way higher standard than clippy lints, they must be almost always right and also not be opinionated. For unused code, it's already a thing that rust warns about it so from an opinionatedness point of view, the lint should be fine (although there was that case in the experience report where the arguments were in weird orders, where it's a taste based question). For being almost always right, I skimmed @ComputerDruid 's PR and almost all cases seemed like the lint improved the situation, or it was a clear case for the `?` kleene operator.

**Comment by est31** [comment]
Another data point: [syn's changes after the PR was merged](https://github.com/dtolnay/syn/compare/1.0.93...1.0.94).

**Comment by est31** [comment]
More data:

* https://github.com/hermitcore/libhermit-rs/commit/0c19418a15038bde5404b265d0abe2e38ef8ebe1 added an allow
* https://github.com/ComunidadAylas/PackSquash/commit/fb5012ad72dc36b97b730f65072969a4ab79409e removed the rule
* https://github.com/Alexhuszagh/rust-lexical/commit/5c2732c00f998b529e049f578a1566098ba26df3 added an allow (next to an allow for `unused_macros`)

So *most* of the time people are adding allows instead of removing rules. In addition to the list found through github's global commit search, I've also encountered the lint in [glium](https://github.com/glium/glium/commit/f4117f262da4e3d86654b705600b3edb7e71d82b). I've removed the rule there, because it never matched, not even on Rust 1.0 (although glium predates rust 1.0). It also didn't compile any more due to refactors, so I figured it was code added through development that is now not neccessary any more.

Generally I didn't expect to find so much allows compared to the small number of removals. I wonder if the lint is *useful* to people. On the other hand, one of the people who made the commits above added a :heart: reaction to this PR even though they added an allow.

**Comment by est31** [comment]
Crossbeam had the same `compile_error` issue: https://github.com/crossbeam-rs/crossbeam/pull/827

Also a PR that removed unused rules instead of adding an allow.

**Comment by fee1-dead** [comment]
If these rules all unconditionally expand to compile_error, could we do a "dummy expand" to detect such cases? Although it might be too expensive and unnecessarily complex.

**Comment by est31** [comment]
@fee1-dead a dummy expand would add another entire mode to expansion which is quite an undertaking... and it's not fully known how the code block would be expanded as well, as the macro rule might take some input and the input might change whether compile_error is invoked or not.

```Rust
    (@parse $($tail:tt)*) => {
        invoke_mac_if_foo!(compile_error, $($tail:tt)*)
    };
```

And `invoke_mac_if_foo` invokes the given macro if the tail passed to it matches `"foo"`, or something, otherwise it doesn't invoke it.

Even if you "just" wanted to parse the output, you'd be in trouble because of the potential $ syntax that is not normal Rust code. Also you don't know in which context you are.

The simplest solution, the one I'd suggest, is to just scan all tokens in the rule, then look for a literal named `compile_error` followed by a bang. This doesn't catch a lot of stuff, doesn't work with reexports of `compile_error` for example. It also "matches" in the case of the `compile_error` macro being shadowed, which it probably shouldn't. But it's better than nothing.

Due to the highly dynamic, highly untyped nature of Rust's macros I doubt you can get something better than that.

I'd say that 99% of cases just invoke `compile_error` unconditionally and at the top level of the block.

**Pull request merged**
